### PR TITLE
Refactor authentication to address remaining login issues

### PR DESCRIPTION
**Pull request by erickpintor**
Major changes:
- Moves authentication model to `authentication/index.js`
- Moves secret form to authentication modules
- Refactor secret form to show errors
- Validates endpoint and secret before logging in
- Uses session storage API to persist user credentials
  This provides a good user experience as people can now refresh their browser.
  It is a trade off in which we give up a little bit of security in favor of a good
  user experience. Session storage API is still more secure than local storage or cookies.
- Makes user instance contain its root client for the database
- Moves default error handler to the FaunaDB wrapper module
- Introduces `src/setupTests.js` for global mock utils

### CODE REVIEW

File: src/authentication/index.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+export function logout() {
+  return {
+    type: Actions.LOGOUT
+  }
+}
+
+export function reduceAuthentication(state = null, action) {
+  switch (action.type) {
+    case Actions.LOGIN:
+      saveUserSession(action.user)
```
Comment (marrony) [inline src/authentication/index.js]: considering redux advocate that reducers should be pure functions, is that safe since it has side effects?

Comment (erickpintor) [inline src/authentication/index.js]: It has an IO side effect but it is safe enough considering that it does not affect the application state. 

It's not ideal but I broken the that rule cuz I really want to ensure that login users are saved at the session store and, the most important, session store is clear after logout.

**Comment by marrony** [comment]
Shouldn't `src/setupTests.js` live inside `tests` folder? Is there a reason to put in `src`?

**Comment by erickpintor** [comment]
It's how the project is configured due to the usage of `react-scripts`.

**Comment by erickpintor** [comment]
Yeah. I love them! :)

**Pull request merged**